File: rxbinding-leanback-v17/src/main/java/com/jakewharton/rxbinding2/support/v17/leanback/widget/SearchEditTextKeyboardDismissOnSubscribe.java
package com.jakewharton.rxbinding2.support.v17.leanback.widget;

import android.support.v17.leanback.widget.SearchEditText;
import android.support.v17.leanback.widget.SearchEditText.OnKeyboardDismissListener;
import com.jakewharton.rxbinding2.internal.Notification;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class SearchEditTextKeyboardDismissOnSubscribe extends Observable<Object> {
  final SearchEditText view;

  SearchEditTextKeyboardDismissOnSubscribe(SearchEditText view) {
    this.view = view;
  }

  @Override protected void subscribeActual(final Observer<? super Object> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setOnKeyboardDismissListener(listener);
  }

  static class Listener extends MainThreadDisposable implements OnKeyboardDismissListener {
    final SearchEditText view;
    final Observer<? super Object> observer;

    Listener(SearchEditText view, Observer<? super Object> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onKeyboardDismiss() {
      if (!isDisposed()) {
        observer.onNext(Notification.INSTANCE);
      }
    }

    @Override protected void onDispose() {
      // TODO: set to null once http://b.android.com/187101 is released.
      view.setOnKeyboardDismissListener(new SearchEditText.OnKeyboardDismissListener() {
        @Override
        public void onKeyboardDismiss() {
        }
      });
    }
  }
}

File: rxbinding-leanback/src/main/java/com/jakewharton/rxbinding3/leanback/SearchEditTextKeyboardDismissOnSubscribe.kt
@file:JvmName("RxSearchEditText")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.leanback

import androidx.annotation.CheckResult
import androidx.leanback.widget.SearchEditText
import androidx.leanback.widget.SearchEditText.OnKeyboardDismissListener
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable which emits the keyboard dismiss events from `view`.
 *
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun SearchEditText.keyboardDismisses(): Observable<Any> {
  return SearchEditTextKeyboardDismissOnSubscribe(this)
}

private class SearchEditTextKeyboardDismissOnSubscribe(
  private val view: SearchEditText
) : Observable<Any>() {

  override fun subscribeActual(observer: Observer<in Any>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setOnKeyboardDismissListener(listener)
  }

  private class Listener(
    private val view: SearchEditText,
    private val observer: Observer<in Any>
  ) : MainThreadDisposable(), OnKeyboardDismissListener {

    override fun onKeyboardDismiss() {
      if (!isDisposed) {
        observer.onNext(Unit)
      }
    }

    override fun onDispose() {
      view.setOnKeyboardDismissListener(null)
    }
  }
}
